src/bundle.c

```diff
@@ -2014,6 +2292,134 @@ gboolean load_manifest_from_bundle(RaucBundle *bundle, RaucManifest **manifest,
+ */
+static gboolean prepare_verity(RaucBundle *bundle, gchar *loopname, gchar* mount_point, GError **error)
+{
+	gboolean res = FALSE;
+	GError *ierror = NULL;
+	g_autoptr(GError) ierror_dm = NULL;
+	g_autoptr(RaucDM) dm_verity = r_dm_new_verity();
+
+	g_return_val_if_fail(bundle != NULL, FALSE);
+	g_return_val_if_fail(loopname != NULL, FALSE);
+	g_return_val_if_fail(mount_point != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	res = check_manifest_external(bundle->manifest, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	dm_verity->lower_dev = g_strdup(loopname);
+	dm_verity->data_size = bundle->size - bundle->manifest->bundle_verity_size;
+	dm_verity->root_digest = g_strdup(bundle->manifest->bundle_verity_hash);
+	dm_verity->salt = g_strdup(bundle->manifest->bundle_verity_salt);
+
+	res = r_dm_setup(dm_verity, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	res = r_mount_bundle(dm_verity->upper_dev, mount_point, &ierror);
+
+	if (!r_dm_remove(dm_verity, TRUE, &ierror_dm)) {
+		g_warning("failed to mark dm verity device for removal: %s", ierror_dm->message);
+	}
+
+	if (!res) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+/*
+ * Sets up dm-verity and dm-crypt for reading crypt bundles.
+ *
+ * [ /dev/dm-x ]
+ *      🠗
+ * [ dm-crypt  ]
+ *      🠗
+ * [ dm-verity ]
+ *      🠗
+ * [  bundle   ]
+ */
+static gboolean prepare_crypt(RaucBundle *bundle, gchar *loopname, gchar* mount_point, GError **error)
+{
+	gboolean res = FALSE;
+	GError *ierror = NULL;
+	g_autoptr(GError) ierror_dm = NULL;
+	g_autoptr(RaucDM) dm_verity = r_dm_new_verity();
+	g_autoptr(RaucDM) dm_crypt = r_dm_new_crypt();
+
+	g_return_val_if_fail(bundle != NULL, FALSE);
+	g_return_val_if_fail(loopname != NULL, FALSE);
+	g_return_val_if_fail(mount_point != NULL, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	res = check_manifest_external(bundle->manifest, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	/* setup dm-verity */
+	dm_verity->lower_dev = g_strdup(loopname);
+	dm_verity->data_size = bundle->size - bundle->manifest->bundle_verity_size;
+	dm_verity->root_digest = g_strdup(bundle->manifest->bundle_verity_hash);
+	dm_verity->salt = g_strdup(bundle->manifest->bundle_verity_salt);
+
+	res = r_dm_setup(dm_verity, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		return FALSE;
+	}
+
+	/* setup dm-crypt */
+	dm_crypt->lower_dev = g_strdup(dm_verity->upper_dev);
+	dm_crypt->data_size = bundle->size - bundle->manifest->bundle_verity_size;
+	dm_crypt->key = g_strdup(bundle->manifest->bundle_crypt_key);
+
+	res = r_dm_setup(dm_crypt, &ierror);
+	if (!res) {
+		/* ensure the already-set-up dm-verity layer is cleaned */
+		if (!r_dm_remove(dm_verity, TRUE, &ierror_dm)) {
+			g_warning("Failed to mark dm-verity device for removal: %s", ierror_dm->message);
+		}
```

Comment:
As this is the error handling, we should not use deferred cleanup, so we're not marking for removal.

Reply:
Oh, indeed. Will change to `FALSE`.

Reply:
The warning also needs to be adjusted.

Reply:
gnah, right